doc/neps/nep-0030-duck-array-protocol.rst

```diff
@@ -90,15 +91,20 @@ a complete implementation would look like the following:
             return self
 
         def __array__(self):
-            return TypeError
+            return TypeError("NumPyLikeArray can not be converted to a numpy array. "
+                             "You may want to use np.duckarray.")
 
 The implementation above exemplifies the simplest case, but the overall idea
 is that libraries will implement a ``__duckarray__`` method that returns the
-original object, and ``__array__`` solely for the purpose of raising a
-``TypeError``, thus preventing unintentional NumPy-coercion. In case of existing
-libraries that don't already implement ``__array__`` but would like to use duck
-array typing, it is advised that they they introduce both ``__array__`` and
-``__duckarray__`` methods.
+original object, and an ``__array__`` method that either creates and returns an
+appropriate NumPy array, or raises a``TypeError`` to prevent unintentional use
+as an object in a numpy array (if ``np.asarray`` is called on an arbitrary
+object that does not imipiment ``__array__()``, it will create a numpy object
```

Comment:
```suggestion
object that does not implement ``__array__``, it will create a NumPy array).
```

Reply:
Actually, I think it's helpful to be clear that it creates a single NumPy scalar with the object dtype:

```
In [3]: np.asarray(ArbitraryClass())
Out[3]: array(<__main__.ArbitraryClass object at 0x10377c208>, dtype=object)
```
This is very much not what people would expect or want if ArbitraryClass is thought to be array-ish.

Maybe there's a better way to word that?

Reply:
Since the returned type is `array`, I would say that is a de-facto array. In my understanding, we would call an element of the array a scalar, or a single value such as `np.int32(0)`, both would be what I understand as an array scalar (i.e., a scalar that is a subset/element of an array).

Reply:
Well, not exactly. NumPy has an "Scalar" object, somewhat different than a size-1 array. YOu can tell you have one because its shape is ().

When you pass a class that asarray can't figure out how to turn into an array, it creates such a numpy scalar, of dtype object -- which is very unlikely to be what the user wants.

```
In [4]: arr = np.asarray(ArbitraryClass())

In [5]: arr.dtype
Out[5]: dtype('O')

In [6]: arr.shape
Out[6]: ()

In [7]: arr
Out[7]: array(<__main__.ArbitraryClass object at 0x10b4c7a58>, dtype=object)
```

Reply:
Ok, you're right. I normally see that referred to as a 0-dimensional array, not array scalar. But according to [this](https://docs.scipy.org/doc/numpy-1.15.0/reference/arrays.ndarray.html#calculation), the name you chose is also perfectly valid:

> ... This behavior is also the default if self is a 0-dimensional array or array scalar. (An array scalar is an instance of the types/classes float32, float64, etc., whereas a 0-dimensional array is an ndarray instance containing precisely one array scalar.)

That said, I agree with your name you originally proposed, let's keep that. Thanks for engaging here!